ament_cmake_auto/cmake/ament_auto_find_build_dependencies.cmake

```diff
@@ -27,18 +32,40 @@
 # @public
 #
 macro(ament_auto_find_build_dependencies)
-  if(ARGN)
+  cmake_parse_arguments(_ARG "" "" "REQUIRED" ${ARGN})
+  if(_ARG_UNPARSED_ARGUMENTS)
     message(FATAL_ERROR "ament_auto_find_build_dependencies() called with "
-      "unused arguments: ${ARGN}")
+      "unused arguments: ${_ARG_UNPARSED_ARGUMENTS}")
   endif()
 
   if(NOT _AMENT_PACKAGE_NAME)
     ament_package_xml()
   endif()
 
+  # ensure that the caller isn't expecting additional packages to be found
+  set(_build_and_buildtool_depends
+    ${${PROJECT_NAME}_BUILD_DEPENDS}
+    ${${PROJECT_NAME}_BUILDTOOL_DEPENDS}
+  )
+  set(_ignored_packages "")
+  foreach(_package_name ${_ARG_REQUIRED})
+    if(NOT _package_name IN_LIST _build_and_buildtool_depends)
+      list(APPEND _ignored_packages ${_package_name})
+    endif()
+  endforeach()
+  if(_ignored_packages)
+    message(FATAL_ERROR "ament_auto_find_build_dependencies() called with "
+      "required packages that are not listed as a build/buildtool dependency in "
+      "the package.xml: ${_ignored_packages}")
+  endif()
+
   # try to find_package() all build dependencies
   foreach(_dep ${${PROJECT_NAME}_BUILD_DEPENDS})
-    find_package(${_dep} QUIET)
+    if(_dep IN_LIST _ARG_REQUIRED)
+      find_package(${_dep} REQUIRED QUIET)
+    else()
+      find_package(${_dep} QUIET)
+    endif()
```

Comment:
In order to avoid duplicating logic I would suggest to write this block as following:

```
set(_REQUIRED_KEYWORD "")
if(_dep IN_LIST _ARG_REQUIRED)
  set(_REQUIRED_KEYWORD "REQUIRED")
endif()
find_package(${_dep} QUIET ${_REQUIRED_KEYWORD})
```